linux/devdoc/socket_transport_linux.md

```diff
@@ -0,0 +1,284 @@
+
+If `recv` a value less then 0, `socket_transport_receive` shall do the following:
+
+- If `errno` is `EAGAIN` or `EWOULDBLOCK`, `socket_transport_receive` shall break out of loop and return `SOCKET_RECEIVE_WOULD_BLOCK`.
+
+- If `errno` is `ECONNRESET`, `socket_transport_receive` shall break out of the looop and return `SOCKET_RECEIVE_SHUTDOWN`.
+
+- else `socket_transport_receive` shall break out of the looop and return `SOCKET_RECEIVE_ERROR`.
+
+If `recv` returns a `0` value, `socket_transport_receive` shall break and return `SOCKET_RECEIVE_SHUTDOWN`.
+
+Else `socket_transport_receive` shall do the following:
+
+- `socket_transport_receive` shall test that the total recv size will not overflow.
+
+- `socket_transport_receive` shall store the received byte size.
+
+If `bytes_recv` is not `NULL`, `socket_transport_send` shall set `bytes_recv` the total bytes received.
+
+`socket_transport_receive` shall call `sm_exec_end`.
+
+### socket_transport_listen
+
+```c
+MOCKABLE_FUNCTION(, int, socket_transport_listen, SOCKET_TRANSPORT_HANDLE, socket_transport, uint16_t, port);
+```
+
+`socket_transport_listen` is to listen for incoming connections.
+
+If `socket_transport` is `NULL`, `socket_transport_listen` shall fail and return a non-zero value.
+
+If `port` is `0`, `socket_transport_listen` shall fail and return a non-zero value.
+
+If the transport type is not `SOCKET_SERVER`, `socket_transport_listen` shall fail and return a non-zero value.
+
+`socket_transport_listen` shall call `sm_open_begin` to begin the open.
+
+If `sm_open_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_listen` shall fail and return a non-zero value.
+
+`socket_transport_listen` shall call `socket` with the params `AF_INET`, `SOCK_STREAM` and `IPPROTO_TCP`.
+
+`socket_transport_listen` shall bind to the socket by calling `bind`.
+
+`socket_transport_listen` shall start listening to incoming connection by calling `listen`.
+
+If successful `socket_transport_listen` shall call `sm_open_end` with `true`.
+
+If any failure is encountered, `socket_transport_listen` shall call `sm_open_end` with `false`, fail and return a non-zero value.
+
+### socket_transport_accept
+
+```c
+MOCKABLE_FUNCTION(, SOCKET_TRANSPORT_HANDLE, socket_transport_accept, SOCKET_TRANSPORT_HANDLE, socket_transport);
+```
+
+`socket_transport_accept` accepts the incoming connections.
+
+If `socket_transport` is `NULL`, `socket_transport_accept` shall fail and return a non-zero value.
+
+If the transport type is not `SOCKET_SERVER`, `socket_transport_accept` shall fail and return a non-zero value.
+
+`socket_transport_accept` shall call `sm_exec_begin`.
+
+If `sm_exec_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_accept` shall fail and return `SOCKET_SEND_ERROR`.
+
+`socket_transport_accept` shall call `accept` to accept the incoming socket connection.
+
+`socket_transport_accept` shall set the incoming socket to non-blocking.
```

Comment:
> set the incoming socket to non-blocking [](http://example.com/codeflow?start=31&length=40)

can it be set after the call to accept?

Reply:
Yes, it has to be set after accept because the socket get returned in the accept call.